KAFKA-14247: Consumer background thread base implementation

philipnee opened the pull request:
1-pager: https://cwiki.apache.org/confluence/display/KAFKA/Proposal%3A+Consumer+Threading+Model+Refactor

EventHandler runs a task to ingest the ApplicationEvents.  The task runs on a background thread, and consumes events from the ApplicationEventQueue, and produces events to the BackgroundEventQueue.

The PR consist of the basic skeleton of the background thread with
- network client
- basic demonstration of the consumption logic

Continuation of https://github.com/apache/kafka/pull/12663

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

philipnee committed:
Basic event handler definition

philipnee committed:
Define EventHandler interface

philipnee committed:
background thread

background thread

philipnee committed:
lint error

philipnee committed:
moving packages

philipnee committed:
reduce the scope

philipnee committed:
PR comments for naming and documentation

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerBackgroundThread.java:
```diff
@@ -0,0 +1,78 @@
+ *  <li>Executing the rebalance flow.</li>
+ * </ul>
+ */
+public class ConsumerBackgroundThread extends KafkaThread implements AutoCloseable {
```
IIRC, `AutoCloseable` is usually for cases of "try-with-resources." Is that how it will be used?

philipnee replied:
good call, I think probably not? I think we just want to start the thread.  I'm actually not sure why I added that there.

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerBackgroundThread.java:
```diff
@@ -0,0 +1,78 @@
+
+    private final BlockingQueue<ConsumerRequestEvent> consumerRequestEvent;
+    private final BlockingQueue<ConsumerResponseEvent> consumerResponseEvent;
+    private Time time;
```
Nit-picky, but this can and probably should be `final` too.

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerBackgroundThread.java:
```diff
@@ -0,0 +1,78 @@
+    // control variables
+    private volatile boolean closed = false;
+
+    public ConsumerBackgroundThread(BlockingQueue<ConsumerRequestEvent> consumerRequestEvents,
```
The constuctor logic seems a bit reversed from what I'd expect. 

I would imagine the three-arg version would be be:

```
super(CONSUMER_BACKGROUND_THREAD_PREFIX, true);
this.time = time;
this.consumerRequestEvent = consumerRequestEvents;
this.consumerResponseEvent = consumerResponseEvents;
```

And then the two-arg version simply be:

```
this(Time.SYSTEM, consumerRequestEvents, consumerResponseEvents);
```

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerBackgroundThread.java:
```diff
@@ -0,0 +1,78 @@
+    public void run() {
+        try {
+            while(!closed) {
+                // TODO: implementation will be added here
```
Magic happens here... 😄

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultEventHandler.java:
```diff
@@ -0,0 +1,49 @@
+ * to be consumed by the background thread and poll {@linkBackgroundEvent} produced by the background thread.
+ */
+public class DefaultEventHandler implements EventHandler {
+    private BlockingQueue<ApplicationEvent> applicationEvents;
```
These two `BlockingQueue`s can be `final`?

philipnee replied:
made the change on the other PR.

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/EventHandler.java:
```diff
@@ -0,0 +1,27 @@
+ */
+public interface EventHandler {
+    public BackgroundEvent poll();
+    public boolean add(ApplicationEvent event);
```
Can you describe that the `boolean` that is returned from `add` represents?

philipnee replied:
javadoc added to https://github.com/apache/kafka/pull/12663

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultEventHandler.java:
```diff
@@ -0,0 +1,49 @@
+
+    @Override
+    public BackgroundEvent poll() {
+        return backgroundEvents.poll();
```
This can return `null`, right?

philipnee replied:
I wrapped it with optional ...see https://github.com/apache/kafka/pull/12663

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultEventHandler.java:
```diff
@@ -0,0 +1,49 @@
+
+    @Override
+    public boolean add(ApplicationEvent event) {
+        return applicationEvents.add(event);
```
According to the [JavaDoc for `AbstractQueue`](https://docs.oracle.com/javase/8/docs/api/java/util/AbstractQueue.html#add-E-):

> This implementation returns `true` if `offer` (sic) succeeds, else throws an `IllegalStateException`.

Do we want an exception thrown if the buffer is full? Or do we want the [`offer`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/LinkedBlockingQueue.html#offer-E-) method which returns `false`?

> Inserts the specified element at the tail of this queue if it is possible to do so immediately without exceeding the queue's capacity, returning `true` upon success and `false` if this queue is full. When using a capacity-restricted queue, this method is generally preferable to method [`add`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/BlockingQueue.html#add-E-), which can fail to insert an element only by throwing an exception.

philipnee committed:
Documentation on the beahvior

Documentation on the beahvior

kirktrue reviewed (changes requested):
Are `ConsumerRequestEvent`/`ConsumerResponseEvent` and `ApplicationEvent`/`BackgroundEvent` the same thing?

philipnee committed:
Add isEmpty and remove excessive public

philipnee committed:
Remove unused import

philipnee committed:
Handle capacity limitation

philipnee committed:
clean up a comment

philipnee committed:
A stubbed event handler to demonstrate the usage.

Stubbed event handler

philipnee committed:
typo

philipnee committed:
Revert "typo"

This reverts commit 625b16c7acf42ccced1ca86cfcbb054b99fa7e59.

philipnee committed:
Revert "A stubbed event handler to demonstrate the usage."

This reverts commit 816b285b6e1a5112be8ce97a000b9a2a14bf7020.

philipnee committed:
Prototyping

philipnee committed:
clean up and documentation

clean up

documentation

implemented poll and commitSync

clean up

clean up

philipnee committed:
Testing, background thread impl.

initialize backgorund thread.

refactor stuff a bit

Added unit tests

more unit tests

renaming and documentation

clean up

clean up

clean up

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultEventHandler.java:
```diff
@@ -0,0 +1,76 @@
+ * An {@code EventHandler} that uses a single background thread to consume {@code ApplicationEvent} and produce
+ * {@code BackgroundEvent} from the {@ConsumerBackgroundThread}.
+ */
+public class DefaultEventHandler implements EventHandler {
```
probably not worth making this an abstract class.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerBackgroundThread.java:
```diff
@@ -0,0 +1,206 @@
+ * in this state.</li>
+ * </ul>
+ */
+public class ConsumerBackgroundThread extends KafkaThread {
```
Probably not worth making this an abstract class.

philipnee committed:
renaming for clarity

philipnee committed:
fixes

philipnee committed:
Refactor the background to make clear it is a network io thread

network IO thread

philipnee committed:
wip

philipnee committed:
delete file

philipnee committed:
tests

philipnee committed:
extra space

philipnee committed:
spaces

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/BackgroundThreadRunnable.java:
```diff
@@ -0,0 +1,26 @@
+import java.io.Closeable;
+
+/**
+ * The {@code EventHandler} constructs a thread that runs {@code BackgroundThreadRunnable} to handle network requests
```
Using `{@link EventHander}` is a little nicer when navigating via JavaDoc, FWIW.

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThreadRunnable.java:
```diff
@@ -0,0 +1,235 @@
+    private long retryBackoffMs;
+    private int heartbeatIntervalMs;
+    private boolean running = false;
+    private Optional<ApplicationEvent> inflightEvent;
```
There can be more than one inflight event, right? Or is it intentional to block and wait for the next event to be processed _completely_ before pulling anything else off the queue?

philipnee replied:
yes, for now, I think we stay with 1 inflight event.

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThreadRunnable.java:
```diff
@@ -0,0 +1,235 @@
+    private final ConsumerMetadata metadata;
+    private String clientId;
+    private long retryBackoffMs;
+    private int heartbeatIntervalMs;
```
Super nitpicky, but why is `retryBackoffMs` a `long`, but `heartbeatIntervalMs` is an `int`?

philipnee replied:
yeah... unfortunate it's the type used in several other classes, for example
```
public GroupRebalanceConfig(final int sessionTimeoutMs,
                                final int rebalanceTimeoutMs,
                                final int heartbeatIntervalMs,
                                String groupId,
                                Optional<String> groupInstanceId,
                                long retryBackoffMs,
                                boolean leaveGroupOnClose)
```

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThreadRunnable.java:
```diff
@@ -0,0 +1,235 @@
+        if (this.inflightEvent.isPresent()) {
+            return this.inflightEvent;
+        }
+        return Optional.ofNullable(this.applicationEventQueue.poll());
```
We should use the version of `poll` that takes a timeout, right?

philipnee replied:
i thought about it, but I think maybe we don't want to block here, as each loop is already waiting for a `retryBackoffMs`. I'm thinking, if the queue returns nothing, then maybe we give it another try in the subsequent loop?

guozhangwang replied:
+1 to @philipnee 's comment. The main reason is that I think eventually the background thread would not only try to just take events from the queue, since even though there's no events incoming, it would still try to repeatedly send fetch request and handle fetch responses -- well, for now maybe we cannot accumulate more than one fetched response per partition, but we may lift that with a memory limit in the future --- so the `runOnce` pattern would likely be:

1. try to get some event from the queue, if yes, then handle it and potentially send the corresponding request;
2. if the queue is empty, then try to send a fetch request.
3. poll the network client in order to receive any responses ready-to-read on the socket.

In that case, we would not want to wait on the queue at all if there's none, but directly move on to step 2) in each iteration.

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThreadRunnable.java:
```diff
@@ -0,0 +1,235 @@
+                log.info("consuming a NOOP event");
+                return true;
+            default:
+                inflightEvent = Optional.empty();
```
I think we should keep the places that `inflightEvent` are updated to a minimum. This method returning `false` will result in `pollOnce` resetting the `inflightEvent` to `Optional.empty()` again, right?

philipnee replied:
you are right, we need to reset it upon successful execution or some failure.

philipnee replied:
let me think harder to keep the reset logic minimal.

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThreadRunnable.java:
```diff
@@ -0,0 +1,235 @@
+                return true;
+            default:
+                inflightEvent = Optional.empty();
+                log.info("unsupported event type: {}", event.type);
```
I think this should be a `warn`, not `info`?

philipnee replied:
warn!

philipnee committed:
tests seem flakey

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThreadRunnable.java:
```diff
@@ -0,0 +1,235 @@
+    }
+
+    // VisibleForTesting
+    DefaultBackgroundThreadRunnable(Time time,
```
Should this constructor call the "main" constructor? If not, it should call `setConfig` too, right?

philipnee replied:
yap

philipnee replied:
This constructor is meant for testing because a few of the clients are actually need to be initiated here locally, e.g. metadata, network client.

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThreadRunnable.java:
```diff
@@ -0,0 +1,235 @@
+        this.metrics = new Metrics();
+    }
+
+    private void setConfig() {
```
Could the contents of `setConfig` be moved into the "main" constructor? Then you wouldn't need to hold the `ConsumerConfig` any more.

philipnee replied:
good call.

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThreadRunnable.java:
```diff
@@ -0,0 +1,235 @@
+    @Override
+    public void close() {
+        this.running = false;
+        Utils.closeQuietly(networkClient, "consumer network client");
```
I believe the `Metrics` and `ConsumerMetadata` classes have a `close()` method that should be invoked here too.

philipnee replied:
I wonder if the metrics should be closed by the polling thread as it is also used there.

kirktrue reviewed (changes requested):
I am finding it a little hard to follow the changes because it's not a self-contained PR. So it's hard to keep the former PRs in mind when reviewing subsequent PRs 😟

philipnee committed:
PR comments

philipnee committed:
cleaned up PR

blank line

reduce logic

fixed a few tests

documentation

fix the logic

wip

philipnee committed:
exception handling

philipnee committed:
clean up unused import

philipnee committed:
More documentation and better exception handling

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/BackgroundThreadRunnable.java:
```diff
@@ -0,0 +1,25 @@
+/**
+ * Background thread runnable that handles network IO such as fetching and committing.
+ */
+public interface BackgroundThreadRunnable extends Runnable, Closeable {
```
I think the additional value of this interface is small given the already existed `DefaultEventHandler` class. That is, the default implementation of the event handler would always have a background thread anyways, while mocks would not have additional threads.

So I'd suggest we remove this interface and just let `DefaultEventHandler` contains a `BackgroundThread` directly.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThreadRunnable.java:
```diff
@@ -0,0 +1,284 @@
+ * It holds a reference to the {@link SubscriptionState}, which is
+ * initialized by the polling thread.
+ */
+public class DefaultBackgroundThreadRunnable implements BackgroundThreadRunnable {
```
It's probably better to extend the `Thread` class (which implements Runnable anyways) for a couple of reasons: 1) we can set a meaningful thread name, which would be useful when we add metrics for the background activities, as part of the metric tags. 2) The shutting down procedure may be cleaner with the built-in `shutdown and join` mechanisms with the `Thread` class.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThreadRunnable.java:
```diff
@@ -0,0 +1,284 @@
+                fetcherThrottleTimeSensor);
+    }
+
+    public DefaultBackgroundThreadRunnable(Time time,
```
How about extracting the construction of the `networkClient` out of the constructor of the thread, but in the constructor of the `DefaultEventHandler` and then pass in the `networkClient` into the thread? In that way we can save on passing many fields to its constructor.

guozhangwang replied:
Ditto for `ConsumerMetadata`.

guozhangwang replied:
And in this way we would not need two constructors either.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThreadRunnable.java:
```diff
@@ -0,0 +1,284 @@
+            log.debug("{} started", getClass());
+            while (running) {
+                runOnce();
+                time.sleep(retryBackoffMs);
```
Instead of setting a sleep here, I think it's better to have a `waitOnEmptyInputQueue` function with a condition. Where the caller thread would notify when enqueuing a new item.

philipnee replied:
@guozhangwang - I thought we wanted to maintain an active loop because we want to keep sending fetches and rebalance requests despite an empty ApplicationQueue.

guozhangwang replied:
Yeah I think so -- please see my other comment below: https://github.com/apache/kafka/pull/12672#discussion_r990528075. But I still feel that instead of having a sleep for each iteration, we should just consider poll on the network client with a timeout, i.e. supposing our loop would look like:

1. try to get some event from the queue, if yes, then handle it and potentially send the corresponding request;
2. if the queue is empty, then try to send a fetch request.
3. poll the network client in order to receive any responses ready-to-read on the socket.

Then our logic could be: if there's no new actions taken at step 1/2), i.e. we do not have any new items from the queue, and we do not yet need to send any new fetch/rebalance-related requests, then at step 3) we poll for a bit longer time until being notified by the caller that there's new items in the queue; otherwise, at step 3) we just poll without timeout and then immediately move on to the next iteration.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThreadRunnable.java:
```diff
@@ -0,0 +1,284 @@
+        return this.running;
+    }
+
+    public Optional<RuntimeException> exception() {
```
This one seems not used?

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThreadRunnable.java:
```diff
@@ -0,0 +1,284 @@
+                time.sleep(retryBackoffMs);
+            }
+        } catch (InterruptException e) {
+            log.error("The background thread has been interrupted");
```
Why we want to retain interrupt exceptions as well, instead of just ignoring it and continue the next iteration?

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThreadRunnable.java:
```diff
@@ -0,0 +1,284 @@
+     * for demonstration purpose.
+     * @param event a {@link NoopApplicationEvent}
+     */
+    private void process(NoopApplicationEvent event) {
```
This seems like what a mock client needs to do, why we need to add it to the default implementation? Or are we going to remove it in this class later?

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultEventHandler.java:
```diff
@@ -16,45 +16,114 @@
+                metrics,
+                clusterResourceListeners,
+                fetcherThrottleTimeSensor);
+        this.backgroundThread = new KafkaThread("consumer_background_thread", runnable, true);
```
Related to my other comment in the `DefaultBackgroundThreadRunnable`: I think we can let that class to extend `Thread` directly (and then we can also rename it to `BackgroundThread` or `ApplicationEventHandlerThread`)?

philipnee replied:
WDYT: `ConsumerBackgroundThread` to disambiguating itself from other BackgroundThreads

guozhangwang replied:
Yeah that sounds better :)

guozhangwang commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/DefaultEventHandlerTest.java:
```diff
@@ -0,0 +1,135 @@
+        DefaultEventHandler handler = new DefaultEventHandler(
+                time,
+                new ConsumerConfig(properties),
+                aq, bq,
```
nit: newline for new param.

guozhangwang commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/DefaultEventHandlerTest.java:
```diff
@@ -0,0 +1,135 @@
+        BlockingQueue<ApplicationEvent> applicationEventQueue = new LinkedBlockingQueue<>();
+        BlockingQueue<BackgroundEvent> backgroundEventQueue = new LinkedBlockingQueue<>();
+        EventHandler eventHandler = new DefaultEventHandler(
+                new RunOnceBackgroundThreadRunnable(applicationEventQueue, backgroundEventQueue),
```
It's unclear to me why we would need a separate background thread to test this functionality? I think just using the default thread would be sufficient still?

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/NoopApplicationEvent.java:
```diff
@@ -0,0 +1,34 @@
+/**
+ * The event is NoOp. This is intentionally left here for demonstration purpose.
+ */
+public class NoopApplicationEvent extends ApplicationEvent {
```
I think we should move this class to the test package instead of the main package to separate it from other production, non-testing code.

guozhangwang commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/DefaultEventHandlerTest.java:
```diff
@@ -0,0 +1,135 @@
+                applicationEventQueue,
+                backgroundEventQueue);
+        assertTrue(eventHandler.add(new NoopApplicationEvent("hello-world")));
+        while (eventHandler.isEmpty()) { }
```
This is a general comment: I think it worth adding a `poll(timeout)` in the `EventHandler` interface since 1) for testing, we do not need the while loop any more, and 2) in production code, for the caller thread scenarios I think there are cases where a `poll(timeout)` would be useful as well.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultEventHandler.java:
```diff
@@ -16,45 +16,114 @@
+    }
+
+    // VisibleForTesting
+    DefaultEventHandler(BackgroundThreadRunnable runnable,
```
I had a similar thought about those testing-used constructors as the background thread class: for this one, it seems to me that just using the default thread is also okay (see my other comment in the test class); for the one in line 80, I think it's better to let it's caller, a.k.a. KafkaConsumer to construct the aq and bq and then pass them into the constructors, in which case we only need that constructor, and not the one in line 45 any more.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEvent.java:
```diff
@@ -20,4 +20,14 @@
+
+    public enum EventType {
+        COMMIT,
+        NOOP,
```
Ditto, I think we should keep the testing-related element out of the production code eventually. If we are going to remove it later then that's fine.

philipnee replied:
Thanks, I think the NOOP event will be moved to the test package in the future PR; I left it here to elucidate the purpose of a few methods in the background thread.

guozhangwang replied:
Thanks for clarifying this, makes sense.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/NoopBackgroundEvent.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * Noop event. Intentionally left it here for demonstration purpose.
+ */
+public class NoopBackgroundEvent extends BackgroundEvent {
```
Ditto here, I think we should move it to test package, not main package.

guozhangwang commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThreadRunnableTest.java:
```diff
@@ -0,0 +1,85 @@
+        assertFalse(client.active());
+    }
+
+    private DefaultBackgroundThreadRunnable setupMockHandler() {
```
Could we add test coverage here for: 1) it will poll network client (We can mock the network client) in each iteration; 2) it will try to poll from the aq in each iteration.

guozhangwang reviewed (commented):
Thanks @philipnee for the PR. I made a first pass and left some comments.

philipnee committed:
refactor based on PR comment

refactor PR comment

philipnee committed:
Move network client construction to the handler

philipnee committed:
clean up

philipnee committed:
wakeup logic

philipnee committed:
documentation

philipnee committed:
swallow interrupt exception

philipnee committed:
test swallowed exception

philipnee commented:
Hey @guozhangwang , much thanks for the detail reviews, I tried to address some of the comments, please review them.  In particular:
1. poll timeout logic
2. InterruptException handling (A side question here, when the BT got interrupted, shouldn't we terminate instead of swallowing the exception?)
3. Added a couple of tests to test the network client calls in a single poll.
4. I also removed the interface, originally I thought it would be helpful to write stubbed tests.

I left the NOOP event there, but I agree we should move it once we've got an actual event implemented, which should happen soon.

philipnee committed:
remove the test for testing

philipnee committed:
wakeup exception test

philipnee commented:
cc John - @vvcephei

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/EventHandler.java:
```diff
@@ -30,6 +30,14 @@ public interface EventHandler {
+     * @return an Optional of {@link BackgroundEvent} if the value is present
+     * . Otherwise, an empty Optional.
+     */
+    Optional<BackgroundEvent> poll(Duration timeout);
```
What's the case where we'd want to use the non-timeout-based `poll` method?

philipnee replied:
not at this moment, we can implement this once the use case is clear.

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java:
```diff
@@ -0,0 +1,226 @@
+    @Override
+    public void run() {
+        try {
+            log.debug("{} started", getClass());
```
Printing the class name here is a bit redundant since it will be part of the log message (assuming the user doesn't change the log format).

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java:
```diff
@@ -0,0 +1,226 @@
+                try {
+                    runOnce();
+                } catch (WakeupException | InterruptException e) {
+                    log.debug("Exception thrown, background thread won't " +
```
Can you add some comments (in the code) as to why we're handling the exceptions from `runOnce` this way?

philipnee replied:
Sounds good, the main reason is to swallow the Wakeup calls

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java:
```diff
@@ -0,0 +1,226 @@
+            log.error("The background thread failed due to unexpected error",
+                    t);
+            if (t instanceof RuntimeException)
+                this.exception.set(Optional.of((RuntimeException) t));
```
Do we want to use `KafkaException` instead of `RuntimeException`?

philipnee replied:
We could, there are other places in the code that uses RTE so trying to be consistent with the existing code.

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java:
```diff
@@ -0,0 +1,226 @@
+     */
+    void runOnce() {
+        this.inflightEvent = maybePollEvent();
+        log.debug("processing application event: {}", this.inflightEvent);
```
Do we want this logged when the inflight event is `Optional.EMPTY`?

philipnee replied:
hmm good call, probably not.

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java:
```diff
@@ -0,0 +1,226 @@
+     * @param event an {@link ApplicationEvent}
+     * @return true when successfully consumed the event.
+     */
+    private boolean maybeConsumeInflightEvent(ApplicationEvent event) {
```
What's the benefit of having an `EventType` in addition to the `ApplicationEvent`? The type is in the incoming `ApplicationEvent` which appears to be then used in the `switch` statement to cast to the desired `ApplicationEvent` type. Why not just have add an abstract `process` method on the `ApplicationEvent` that does whatever needs to be done?

kirktrue replied:
If we let the `ApplicationEvent` perform whatever logic is needed, that frees up the `DefaultBackgroundThread` from having to encapsulate all the logic of all the events that can happen. Keeping the `ApplicationEvent` and the `DefaultBackgroundThread` a little more loosely-coupled may allow easier unit testing 🤷‍♂️

kirktrue replied:
Then you'd also not need to introduce the `NOOP` bits into this code at all.

philipnee replied:
great idea

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java:
```diff
@@ -0,0 +1,226 @@
+    }
+
+    public void close() {
+        this.wakeup();
```
Does the order of these two lines (`wakeup` and `running = false`) want to be swapped?

`close` is called by the polling thread, right? If `wakeup` is called, it will throw an exception, but then the background thread might potentially loop back around to `runOnce` before the call to set `running` to `false` by the polling thread.

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java:
```diff
@@ -0,0 +1,226 @@
+        this.wakeup();
+        this.running = false;
+        Utils.closeQuietly(networkClient, "consumer network client");
+        Utils.closeQuietly(metadata, "consumer network client");
```
The string should be `"consumer metadata"` (or something) when closing the `metadata` instead, right?

philipnee replied:
thanks... copy paste error

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultEventHandler.java:
```diff
@@ -16,45 +16,185 @@
+                               Sensor fetcherThrottleTimeSensor) {
+        this.applicationEventQueue = applicationEventQueue;
+        this.backgroundEventQueue = backgroundEventQueue;
+        ConsumerMetadata metadata = bootstrapMetadata(logContext,
```
The call to `bootstrapMetadata` is blocking, right? Is that what we want? If so, can you add some comments (in the code) as to that behavior?

philipnee replied:
It should be fine for this call to block, I think because all it does is create a Node object with id = -1 and an IP address. The actual metadata refresh happens during the background thread loop, not during this function call, since there are no network response deps to this call, so it shouldn't be blocking for too long.

kirktrue reviewed (changes requested):
I did another review of the code. I haven't gotten as far as the tests yet.

philipnee committed:
refactor based on PR comment

philipnee committed:
debug message

vvcephei committed:
apply AK code conventions

vvcephei commented:
Hey, @philipnee, @guozhangwang isn't available for reviews right now, so I'll pick it up.

I reviewed it, and it all looks good to me. I had a large number of code convention corrections, so I just went ahead and applied the changes, rather than peppering you with nitpicks. I hope that's ok.

Generally:
* please use `final` when it can be used
* indent 4 spaces instead of 8
* when a method call has too many args to fit within the line length limit, please break it up into one-arg-per-line, rather than just breaking the args at an arbitrary point.

I know that's all arbitrary, and that there are merits to doing things other ways; I'm just trying to keep these files in line with the rest of AK.

Thanks! Once the tests pass, I'll go ahead and merge.

philipnee commented:
Thanks @vvcephei Will be better at the basic styling requirements in the future...
